File: BubbleSort/BubbleSort.cpp
//
// Created by Joseph Islam on 7/14/19.
//

#include <iostream>
#include <array>
#include <vector>

#include <fstream>

using namespace std;

static void BubbleSort(vector<int> &data){
    // it takes data.size() - 1 steps to move the smallest element at the back to the front of the vector
    // Loop data.size() - 1 times
    for (int start = 0; start < data.size() - 1; start++){
        // The greatest element is always at the back after each iteration
        // move through the list from front to back, moving big elements to the back of the list
        for (int target = 0; target < data.size() - 1 - start; target++){
            // Swap if a big element is before a little element
            if (data[target] > data[target+1]){
                // Swap without using a new variable.
                data[target] = data[target] + data[target+1];
                data[target+1] = data[target] - data[target+1];
                data[target] = data[target] - data[target+1];
            }
        }
    }
}

// Reads files in format: str\nint\nint\nint... format
// And returns data in a vector
static vector<int> read_dataVec(const string& filename){
    // Create a vector to store the working data in,
    vector<int> data = vector<int>();
    //  a file to read from,
    ifstream csv(filename);
    // and a line store data in.
    string line;

    // Read through the file on opening
    if (csv.is_open()){
        // Storing each newline in line
        while (getline(csv, line)){
            // And storing potential ints in data.
            try{
                data.push_back(stoi(line));
            }
            catch (invalid_argument e){}
        }
        // Close the file when finished
        csv.close();
    } else{
        // If the file is not open, say so, and yield error code 11
        cout << "Unable to open file";
    }

    // Return data
    return data;
}

int main(int argc, char* argv[]){
    // Make sure user supplies vector filename and vector filename alone
    if (argc != 2){
        printf("Usage: ./main [dataVectorFilename]");
        return 0;
    }

    // Get filename to read data from from commandline args
    string filename = argv[1];

    // Get the data out of that file
    vector<int> data = read_dataVec(filename);

    // And sort it with Bubble Sort
    BubbleSort(data);

    // Finally, print the sorted data
    for (int element: data){
        cout << element << ", ";
    }

    // And return 1 for successful completion
    return 1;
}

File: BubbleSort/BubbleSort.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;

public class Sorts {

    private static void BubbleSort(List<Integer> data) {
        // it takes data.size() - 1 steps to move the smallest element at the back to the front of the vector
        // Loop data.size() - 1 times
        for (int start = 0; start < data.size() - 1; start++) {
            // The greatest element is always at the back after each iteration
            // move through the list from front to back, moving big elements to the front of the list
            for (int target = 0; target < data.size() - 1 - start; target++) {
                // Swap if a big element is before a little element
                if (data.get(target) > data.get(target + 1)) {
                    // Swap without using a new variable to save space.
                    data.set(target, data.get(target) + data.get(target + 1));
                    data.set(target + 1, data.get(target) - data.get(target + 1));
                    data.set(target,  data.get(target) - data.get(target + 1));
                }
            }
        }
    }

    public static LinkedList<Integer> read_dataVec(String file) throws IOException {
        // Create a reader to read the supplied file,
        BufferedReader reader = new BufferedReader(new FileReader(file));
        // A linkedlist to store the read data in,
        LinkedList<Integer> data = new LinkedList<>();
        // A line to store the current line in,
        String line;
        // And a scanner to scan each line with.
        Scanner scanner;

        // Read all the lines, scanning each one for data seperated by commas
        while ((line=reader.readLine()) != null){
            scanner = new Scanner(line);
            scanner.useDelimiter(",");
            // And store it in the data linkedlist (When integers)
            while (scanner.hasNext()){
                String dat = scanner.next();
                try{
                    data.add(Integer.valueOf(dat));
                }
                catch (NumberFormatException e){
                }
            }
        }
        // Return the read data.
        return data;
    }

    public static void main(String[] args) throws IOException {
        // Make sure user supplies vector filename and vector filename alone
        if (args.length != 1){
            System.out.println("Usage: javac Sorts.java [dataVecFile]");
            return;
        }

        // Read the random integers from the supplied file
        LinkedList<Integer> data = read_dataVec(args[0]);
        // and bubblesort it.
        BubbleSort(data);

        // Finally, print the sorted data
        for (Integer element : data) {
            System.out.print(element + ", ");
        }
    }
}